### PR TITLE
New initializer inputs experimental

### PR DESCRIPTION
**Pull request by grdw**

### CODE REVIEW

File: lib/atlas/graph_values.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    include ActiveModel::Validations
+
+    GRAPH_VALUES_FILENAME = 'graph_values.yml'.freeze
+    VALID_GRAPH_METHODS   = %w(
+      demand
+      share
+      child_share
+      parent_share
+      conversion
+      reserved_fraction
```
Comment (antw) [inline lib/atlas/graph_values.rb]: Are `conversion` and `reserved_fraction` used?

**Comment by grdw** [comment]
> The ETSource branch needs to be updated since it still refers to _setters, and three derived datasets fail to validate.

Yes it does 👍 

> The ETE branch has the old-style graph_values.yml used in the specs.

Actually I'm wondering if we need to anything about ETEngine at all. I mean, testing if these `graph_values.yml `actually loads and works is more Atlas's job anyway.

> The init_methods_to_input ETSource rake task will have to be updated to reflect the new graph_values.yml format. I can take care of that if you wish, but it'll be the beginning of next week at the earliest.

I was thinking of removing this task since merging this + updating ETSource will make the `graph_values` actually useable and there would be no need to convert them to an initializer input.

**Comment by grdw** [comment]
```yaml 
energy_power_supercritical_coal:
   demand: 10000.0
   number_of_units: 3
   output:
     electricity: 0.45
     loss: 0.55
```

That's actually practical. Whitelisting can than happen on a node like:

```
# test.ad

- graph_methods = [output.electricity]
```

**Comment by antw** [comment]
> Whitelisting can than happen on a node like: `- graph_methods = [output.electricity]`

Perhaps we could just whitelist all inputs and/or outputs?

```yaml
- graph_methods = [input, ...]
- graph_methods = [output, ...]
- graph_methods = [input, output, ...]
```

I don't imagine there will be many situations where you want to whitelist only a subset of the input or output slots on a node (i.e. if you want to edit electricity, you also need to edit loss).

**Comment by grdw** [comment]
> It has occurred to me that it might be nicer for slot shares to be part of the node section of graph_values.yml, like so:

@antw done! :)

File: lib/atlas/runner/set_attributes_from_graph_values.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      # Private: set_graph_methods
+      #
+      # el = Atlas element either a Slot, Edge or Node
+      # dataset = an Atlas::Datset::Derived
+      def self.set_graph_methods(dataset, refinery_element, atlas_element = nil)
+        atlas_element = atlas_element || refinery_element.get(:model)
+
+        (dataset.graph_values.for(atlas_element) || {})
+          .each_pair do |method, val|
+            if %w(input output).include?(method)
```
Comment (antw) [inline lib/atlas/runner/set_attributes_from_graph_values.rb]: Small suggestion: This creates three new objects for every initializer method (two strings and an array). You can eliminate that by adding a `# frozen_string_literal: true` comment to the top of the file and using `case`:

```ruby
case method
when 'input', 'output'
  set_graph_methods_to_slot!(refinery_element, method, val)
else
  refinery_element.set(method.to_sym, val)
end
```

File: lib/tasks/auto_whitelist.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  Whitelists all nodes edges based on a sparse graph
+DESC
+
+def add_method(el, method)
+  unless el.graph_methods.include?(method)
+    el.graph_methods << method
+  end
+end
+
+task auto_whitelist: :environment do
```
Comment (antw) [inline lib/tasks/auto_whitelist.rb]: Does this not belong in ETSource?

Comment (grdw) [reply lib/tasks/auto_whitelist.rb]: Yes, it should. I'll add it to the list, stating that ETsource probably needs some TLC first when it comes to it's rake tasks.

**Comment by grdw** [comment]
>  If the other branches are compatible and ready

I think they are. I checked [etsource](https://github.com/quintel/etsource/pull/1454) and [etengine](https://github.com/quintel/etengine/pull/963) and they're good to go.

---

*Noted:*

- There are definitely going to be a lot more changes in ETSource but will do them in the coming weeks (in regards to moving initializer inputs -> initializer methods).
- Also ETLocal / Transformer will get some updates aswell regarding these changes.

**Comment by antw** [comment]
Branches on all three repos have been merged.

Really happy with how this turned out. Great work @grdw! :tada:

**Pull request merged**